modify toggle mode style with DatePickerTheme

MaironLucas opened the pull request:
- Add properties to DatePickerThemeData:
    - `toggleModeStyle` allow customization on _DatePickerModeToggleButton TextStyle;
    - `toggleModeForegroundColor` allow changing the color of both button and Icon with consistency;
- Compatible with existing code by the use of defaults;
- Adjust test to cover new properties

How to customize toggle mode style before:

```dart
  showDatePicker(
    context: context,
    firstDate: DateTime(2025),
    lastDate: DateTime(2026),
    builder:
        (context, child) => Theme(
          data: ThemeData.light().copyWith(
            textTheme: TextTheme(titleSmall: TextStyle(fontSize: 16)),
            datePickerTheme: DatePickerThemeData(
              dayStyle: TextStyle(fontSize: 12),
            ),
          ),
          child: child!,
        ),
  );
```

Now:

```dart
  showDatePicker(
    context: context,
    firstDate: DateTime(2025),
    lastDate: DateTime(2026),
    builder:
        (context, child) => DatePickerTheme(
          data: DatePickerThemeData(
            toggleModeStyle: TextStyle(fontSize: 16),
            dayStyle: TextStyle(fontSize: 12),
          ),
          child: child!,
        ),
  );
```

Ref #163866
Ref #160591

<!--
Thanks for filing a pull request!
Reviewers are typically assigned within a week of filing a request.
To learn more about code review, see our documentation on Tree Hygiene: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
-->

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

MaironLucas commented:
I implemented this new feature before seeing that issue #160591 requested a similar feature. That issue already has a linked PR (#161458), but that PR doesn't seem to have received updates since January 24, so I think it's fair to keep my PR as a valid implementation for both issues. 
Also, I took into account the comments and reviews of that PR.

QuncCccccc commented on packages/flutter/lib/src/material/date_picker_theme.dart:
```diff
@@ -360,6 +362,17 @@ class DatePickerThemeData with Diagnosticable {
+  ///
+  /// The [TextStyle.color] of [toggleModeStyle] is not used, [toggleModeForegroundColor]
+  /// is used instead.
+  final TextStyle? toggleModeStyle;
```
Do you think if `toggleButtonTextStyle` would be more straightforward to understand? :)

MaironLucas replied:
Done

QuncCccccc commented on packages/flutter/lib/src/material/date_picker_theme.dart:
```diff
@@ -360,6 +362,17 @@ class DatePickerThemeData with Diagnosticable {
+  /// Overrides the default color used for text labels and icons of toggle mode button.
+  ///
+  /// This is used instead of the [TextStyle.color] property of [toggleModeStyle].
+  final Color? toggleModeForegroundColor;
```
To fix https://github.com/flutter/flutter/issues/160591, I think we'd better also apply this color to the month pick button. As for naming, I think `subHeaderForegroundColor` would be more straightforward to understand.

MaironLucas replied:
Done

QuncCccccc reviewed (commented):
Thank you for your contribution! Overall this looks good to me! Just left some comments below.

QuncCccccc commented on packages/flutter/lib/src/material/calendar_date_picker.dart:
```diff
@@ -458,12 +459,12 @@ class _DatePickerModeToggleButtonState extends State<_DatePickerModeToggleButton
                               widget.title,
                               overflow: TextOverflow.ellipsis,
-                              style: textTheme.titleSmall?.copyWith(color: controlColor),
+                              style: buttonTextStyle?.apply(color: toggleModeForegroundColor),
```
If we provide a `toggleButtonTextStyle ` with a custom text color in date picker theme, but **don't** provide a `subHeaderForegroundColor` in the date picker theme, this change will cause that the customized text color is not respected, and the text color always defaults to the default `subHeaderForegroundColor `.

For the text color here, ideally, I think we want to firstly check whether theme's toggleButtonTextStyle's color is empty, if not empty, we respect the color value; but if empty, then check whether theme's subheaderForegroundColor is empty, if not empty, text color should respect it; if empty, we should go to defaults.toggleButtonTextStyle, if still empty, we should check `defaults.subHeaderForegroundColor`. Does this make sense?

MaironLucas replied:
Yeahh, it makes sense. I'm gonna implement it!

MaironLucas replied:
Done!  I think the check on defaults.subHeaderForegroundColor is redundant because defaults.toggleButtonTextStyle already gets it.

QuncCccccc reviewed (commented):
Thanks for the update! Left some comments below. Please let me know if there's any questions:)

MaironLucas commented on dev/tools/gen_defaults/lib/date_picker_template.dart:
```diff
@@ -62,6 +62,14 @@ class _${blockName}DefaultsM3 extends DatePickerThemeData {
   Color? get backgroundColor => ${componentColor("md.comp.date-picker.modal.container")};
 
+  @override
+  Color? get subHeaderForegroundColor => ${componentColor("md.comp.date-picker.modal.sub.header.foreground")}.withOpacity(0.60);
```
Okay, that makes sense. This part was a bit confusing to me. So I should use something like `md.comp.date-picker.modal.weekdays.label-text` for `subHeaderForegroundColor` and `md.comp.date-picker.modal.range-selection.month.subhead` for `toggleButtonTextStyle`? Thats it?

MaironLucas replied:
Done!

QuncCccccc replied:
Sorry for the late reply. Yes, these tokens are exported from google internal token generator, we should just pick tokens instead of creating new ones ourselves:)

MaironLucas committed:
feat: sub header style with DatePickerTheme

- Add properties to DatePickerThemeData:
    - `toggleButtonTextStyle` allow customization on _DatePickerModeToggleButton
    TextStyle;
    - `subHeaderForegroundColor` allow changing the color of both button
    and Icons with consistency;
- Compatible with existing code by the use of defaults;
- Adjust test to cover new properties

Ref #163866
Ref #160591

MaironLucas commented on dev/tools/gen_defaults/data/date_picker_modal.json:
```diff
@@ -72,5 +71,7 @@
   "md.comp.date-picker.modal.year-selection.year.unselected.hover.state-layer.color": "onSurfaceVariant",
   "md.comp.date-picker.modal.year-selection.year.unselected.label-text.color": "onSurfaceVariant",
-  "md.comp.date-picker.modal.year-selection.year.unselected.pressed.state-layer.color": "onSurfaceVariant"
+  "md.comp.date-picker.modal.year-selection.year.unselected.pressed.state-layer.color": "onSurfaceVariant",
```
Done!

QuncCccccc reviewed (commented):
This looks great to me! Can we add one more test to verify the `datePickerTheme.toggleButtonTextStyle` can override `datePickerTheme.subHeaderForeground` when they both have some custom color values:)

MaironLucas commented on packages/flutter/lib/src/material/calendar_date_picker.dart:
```diff
@@ -450,9 +448,16 @@ class _DatePickerModeToggleButtonState extends State<_DatePickerModeToggleButton
+    final DatePickerThemeData defaults = DatePickerTheme.defaults(context);
+    final TextStyle? buttonTextStyle =
+        datePickerTheme.toggleButtonTextStyle ?? defaults.toggleButtonTextStyle;
+    final Color? toggleModeForegroundColor =
```
@QuncCccccc About the test, I think I don't know exactly how to make it. Should I write a new test at `date_picker_theme_test` or edit one that already exists to verify this override?

MaironLucas replied:
Hi, @QuncCccccc. How is it going? I pushed the commit with two tests, verifying if the color of toggleButtonTextStyle is respected over subHeaderForegroundColor and also a test verifying if subHeaderForeground is used when toggleButtonTextStyle.color is not defined.

QuncCccccc replied:
The tests look good to me! Could you help fix the conflict and the failures, then we are good to go!

MaironLucas replied:
Hi, nice!!
Well, the conflict is fixed. But about the failures, I am unsure of what to do to fix them. Could you help me? @QuncCccccc

QuncCccccc commented on packages/flutter/lib/src/material/date_picker_theme.dart:
```diff
@@ -373,6 +375,16 @@ class DatePickerThemeData with Diagnosticable {
   /// picker. It defaults to the ambient locale provided by [Localizations].
   final Locale? locale;
 
+  /// Overrides the default text style used for the text of toggle mode button
```
nit: 
```suggestion
  /// Overrides the default text style used for the text of toggle mode button.
```

QuncCccccc reviewed (approved):
LGTM! Thank you for your contribution!

justinmc commented on packages/flutter/test/material/date_picker_theme_test.dart:
```diff
@@ -591,6 +607,13 @@ void main() {
     );
     expect(day24Shape.side.width, datePickerTheme.todayBorder?.width);
 
+    // Test the toggle mode button
```
Nit: Period at the end.

justinmc reviewed (approved):
LGTM :+1:. Thanks for jumping in to add this!
